src/main/java/org/folio/rest/impl/TenantPermsAPI.java

```diff
@@ -35,46 +42,376 @@
-          String err = savePermsRes.cause().getMessage();
-          logger.error(err, savePermsRes.cause());
-          asyncResultHandler.handle(Future.succeededFuture(
-              PostTenantpermissionsResponse.respond400WithTextPlain(err)));
-          return;
-        }
-        asyncResultHandler.handle(Future.succeededFuture(
-            PostTenantpermissionsResponse.respond201WithApplicationJson(entity)));
-      });
+      handlePermLists(entity, vertxContext, tenantId)
+          .onSuccess(v -> asyncResultHandler.handle(Future.succeededFuture(
+              PostTenantpermissionsResponse.respond201WithApplicationJson(entity))))
+          .onFailure(t -> {
+            logger.error(t.getMessage(), t);
+            asyncResultHandler.handle(Future.succeededFuture(
+                PostTenantpermissionsResponse.respond400WithTextPlain(t.getMessage())));
+          });
     } catch (Exception e) {
       logger.error(e.getMessage(), e);
       asyncResultHandler.handle(Future.succeededFuture(
           PostTenantpermissionsResponse.respond500WithTextPlain(e.getMessage())));
     }
   }
 
-  private Future<Void> savePermList(List<Perm> permList, Context vertxContext, String tenantId) {
+  private Future<Permission> addMissingModuleContext(Permission perm, ModuleId moduleId,
+      Context vertxContext, String tenantId) {
+    Promise<Permission> promise = Promise.promise();
+    perm.setModuleName(moduleId.getProduct());
+    SemVer semver = moduleId.getSemVer();
+    perm.setModuleVersion(semver != null ? semver.toString() : null);
+
+    PostgresClient.getInstance(vertxContext.owner(), tenantId).update(TABLE_NAME_PERMS, perm,
+        perm.getId(), updateReply -> {
+          if (updateReply.failed()) {
+            Throwable cause = updateReply.cause();
+            logger.error(cause.getLocalizedMessage(), cause);
+            promise.fail(cause);
+            return;
+          }
+          logger.info(String.format("Added defined context to %s", perm.getPermissionName()));
+          promise.complete(perm);
+        });
+    return promise.future();
+  }
+
+  private Future<List<Permission>> getPermsForModule(ModuleId moduleId, List<OkapiPermission> perms,
+      Context vertxContext, String tenantId) {
+    return getPermsByModule(moduleId, vertxContext, tenantId)
+        .compose(existing -> {
+          if (existing.isEmpty() && perms != null) {
+            // this means either:
+            // A) the first time enabling this module, or
+            // B) the permissions exist but don't yet have the moduleName field
+            List<Permission> ret = new ArrayList<>();
+            List<Future> futures = new ArrayList<>(perms.size());
+            perms.forEach(perm ->
+              futures.add(getModulePermByName(perm.getPermissionName(), null, vertxContext, tenantId)
+                  .compose(dbPerm -> {
+                    if (dbPerm != null && PermissionUtils.equals(perm, dbPerm)) {
```

Comment:
I do not think `PermissionUtils.equals(perm, dbPerm)` is needed. For example, the perm name was same but description was updated. In this case, we have to treat it as the same perm during bootstrap.

Reply:
When I say "bootstrap" I'm referring to a very specific thing.  In OKAPI-947 we made an adjustment to call _tenantPermissions for each module enabled for the tenant whenever mod-permissions is installed/upgraded.  During this, it's possible to accurately identify a permission which was system-defined but lacks the module context because those permissions are identical except for the absence of the module context.  That's what this test does.

Reply:
> For example, the perm name was same but description was updated. In this case, we have to treat it as the same perm during bootstrap
This is detected later and handled as an updated permission

Reply:
> When I say "bootstrap" I'm referring to a very specific thing. In OKAPI-947 we made an adjustment to call _tenantPermissions for each module enabled for the tenant whenever mod-permissions is installed/upgraded. During this, it's possible to accurately identify a permission which was system-defined but lacks the module context because those permissions are identical except for the absence of the module context. That's what this test does.

I see. You might want to add another condition to make sure `dbPerm` has no `moduleName` to cover an edge case for `A) the first time enabling this module`. For example, if two modules define the same perm (very unlikely though). Of course, if `dbPerm` has `moduleName`, you might have to throw an error because it would be a collision.

Reply:
yes that's a good suggestion.  I'll work that in, thanks.

Reply:
OK - got this in w/ test case.  Note that we also have to check that the permission isn't a dummy.

Reply:
Why check for dummy? Would not dummy perm have moduleName as well?